Update sam_bot_description for the new Gazebo

Amronos opened the pull request:
Fixes https://github.com/ros-navigation/docs.nav2.org/issues/608.

- [x] Update DiffDrive plugin for the new Gazebo.
- [x] Update IMU for the new Gazebo.
- [x] Update to use stamped twist messages for `/demo/cmd_vel` 
- [x] Make SDF description
- [x] Finish and test sensor description and bridges
- [x] Make necessary changes to navigation and test it.

Amronos committed:
Update DiffDrive, and IMU for new Gazebo

Update nav2_params.yaml

Update to use stamped twist messages

Add SDF description

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

SteveMacenski commented on sam_bot_description/config/bridge_config.yaml:
```diff
@@ -0,0 +1,33 @@
+---
```
I believe we also add in a Camera and Lidar in the Sensors part of the FirstTime setup guide -- but I see in your other PR that its on your checklist :-) Just leaving the note here that its missing

Amronos replied:
Should be good to go with [57db422](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/57db4224d143056d692990a95f7cc3ad7a7cb45b).

SteveMacenski commented on sam_bot_description/config/bridge_config.yaml:
```diff
@@ -0,0 +1,33 @@
+  ros_type_name: "rosgraph_msgs/msg/Clock"
+  gz_type_name: "gz.msgs.Clock"
+  direction: GZ_TO_ROS
+
```
Do we need TF as well?

Amronos replied:
I think the ekf node should publish the transforms instead of Gazebo. I have documented this [here](https://github.com/ros-navigation/docs.nav2.org/pull/618/files#diff-ec112ab174676a9646815862ab1c29c95529fa7fe6bd2b76bd8292da589b8c8aR276).

SteveMacenski replied:
I guess i haven't typically used EKF when using gazebo, I just let the diff drive plugin provide TF as pretty close to ground truth. I'm open to both ways, but just what I've usually done and makes it a little easier for folks to understand on a tutorial level

Also, it might help in the issues you've mentioned by just removing EKF from the situation :wink:

Amronos replied:
I guess having EKF would be a good idea for a real robot, but not really for Gazebo. I can add the bridge and then maybe make launching EKF an argument of the launch file. Then I can edit the tutorials accordingly.

Amronos replied:
> I guess having EKF would be a good idea for a real robot, but not really for Gazebo. I can add the bridge and then maybe make launching EKF an argument of the launch file. Then I can edit the tutorials accordingly.

Done in [7c1ddc8](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/7c1ddc8ab2ecf3e25d7839637216bf9199a04900).

SteveMacenski commented on sam_bot_description/src/description/sam_bot_description.sdf:
```diff
@@ -0,0 +1,302 @@
+
+      <xacro:box_inertia m="0.1" w="0.1" d="0.1" h="0.1"/>
+
+      <sensor name="imu_sensor" type="imu">
```
I sthis publishing data? I think this needs to be something like 

```
<plugin filename="gz-sim-imu-system"
        name="gz::sim::systems::Imu">
</plugin>
```

Amronos replied:
Fixed with [54b7e2e](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/54b7e2ea85228ef297f8256ecf92971d20cecaef), thanks!

SteveMacenski commented on sam_bot_description/src/description/sam_bot_description.urdf:
```diff
@@ -222,31 +216,33 @@
-        <remap from="cmd_vel" to="/cmd_vel"/>
-      </ros>
-
+    <plugin filename="gz-sim-diff-drive-system" name="gz::sim::systems::DiffDrive">
```
I believe this is in the SDF file now, right?

Amronos replied:
Yes, it's in both the URDF and SDF.

SteveMacenski commented on sam_bot_description/src/description/sam_bot_description.urdf:
```diff
@@ -222,31 +216,33 @@
+      <child_frame_id>base_link</child_frame_id>
+    </plugin>
+
+    <plugin
```
I believe this is in the SDF file now, right?

Amronos replied:
Yes, it's in both the URDF and SDF.

SteveMacenski commented on sam_bot_description/src/description/sam_bot_description.urdf:
```diff
@@ -279,11 +275,13 @@
 
   <gazebo reference="lidar_link">
-    <sensor name="lidar" type="ray">
+    <sensor name="lidar" type="gpu_lidar">
```
I believe this should go into the SDF file now, right? 

Also configured like: https://gazebosim.org/docs/latest/sensors/#lidar-sensor

Amronos replied:
Will be adding this to the SDF file also.

Amronos replied:
Should be good to go with [57db422](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/57db4224d143056d692990a95f7cc3ad7a7cb45b). Also it looks like both ``<sensor><ray></ray></sensor>`` and ``<sensor><lidar></lidar></sensor>`` work, but I changed it to ``<ray>`` anyways.

SteveMacenski commented on sam_bot_description/src/description/sam_bot_description.urdf:
```diff
@@ -350,10 +341,13 @@
 
   <gazebo reference="camera_link">
-    <sensor name="depth_camera" type="depth">
+    <sensor name="depth_camera" type="depth_camera">
```
Ditto with lidar :-)

Amronos replied:
Will be adding this to the SDF file also.

Amronos replied:
Should be good to go with [57db422](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/57db4224d143056d692990a95f7cc3ad7a7cb45b).

SteveMacenski reviewed (commented):
First of all, a huge, huge thank you for taking this on, it really helps alot to get some support in updating documentation :-)

Amronos committed:
Add plugins to world file

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Fix/Update sensors and add them to SDF also

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Update robot footprint

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Update imu_sensor frame_id

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

SteveMacenski commented:
Let me know when you want me to take a look again! I don't want to nitpick mid-update :-)

Amronos commented:
> Let me know when you want me to take a look again! I don't want to nitpick mid-update unless you want me to :-)

Go ahead with it, everything should be good to go.

There is just one problem I have been trying to solve for a while now though:
On running:
```
ros2 run teleop_twist_keyboard teleop_twist_keyboard --ros-args -p stamped:=true --remap cmd_vel:=/demo/cmd_vel
```
It does not seem like the robot moves with every command even though on running ``gz topic -et /demo/cmd_vel`` I can see the messages getting received by Gazebo. Also, the odom->base_link transform doesn't appear to be changing in the same way in ROS & Gazebo. This occurs with and without the ``ekf_node`` (when not using it I bridge the transforms from Gazebo to ROS).

Amronos committed:
Make caster close to frictionless

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos commented:
> There is just one problem I have been trying to solve for a while now though: On running:
> 
> ```
> ros2 run teleop_twist_keyboard teleop_twist_keyboard --ros-args -p stamped:=true --remap cmd_vel:=/demo/cmd_vel
> ```
> 
> It does not seem like the robot moves with every command (happens more on publishing a value on angular z) even though on running `gz topic -et /demo/cmd_vel` I can see the messages are getting received by Gazebo. Also, the odom->base_link transform doesn't appear to be changing in the same way in ROS & Gazebo. This occurs with and without the `ekf_node` (when not using it I bridge the transforms from Gazebo to ROS).

@SteveMacenski this should be fixed with the latest commit. The only problem is that when the imu topic is included for the ``ekf_node``, the robot rotates more on the z-axis than it should. I am not sure if this is a problem of Gazebo or robot_localization.

SteveMacenski commented on sam_bot_description/config/bridge_config.yaml:
```diff
@@ -0,0 +1,65 @@
+  gz_type_name: "gz.msgs.PointCloudPacked"
+  direction: GZ_TO_ROS
+
+- topic_name: "/depth_camera/image"
```
There is an image bridge, maybe would be good to use that? https://github.com/gazebosim/ros_gz/tree/ros2/ros_gz_image I have some examples doing this in some AMD demos https://github.com/open-navigation/opennav_amd_demonstrations/blob/main/honeybee_gazebo/launch/simulation.launch.py#L124-L144. It uses image transport which gives you some more options like using compression

Amronos replied:
Done in [29276cb](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/29276cbcfe4fdf6c89f9f7b312f9e39b67dd0b64)

SteveMacenski commented on sam_bot_description/launch/display.launch.py:
```diff
@@ -1,60 +1,74 @@
 
 def generate_launch_description():
-    pkg_share = launch_ros.substitutions.FindPackageShare(package='sam_bot_description').find('sam_bot_description')
+    pkg_share = FindPackageShare(package='sam_bot_description').find('sam_bot_description')
```
Using https://github.com/ros-navigation/navigation2/blob/main/nav2_bringup/launch/bringup_launch.py#L17 instead lets you remove the `find` bits, it returns an actual value, string path instead of a launch substitution

Amronos replied:
Done in [29276cb](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/29276cbcfe4fdf6c89f9f7b312f9e39b67dd0b64).

SteveMacenski commented on sam_bot_description/launch/display.launch.py:
```diff
@@ -1,60 +1,74 @@
-        launch.actions.ExecuteProcess(cmd=['gazebo', '--verbose', '-s', 'libgazebo_ros_init.so', '-s', 'libgazebo_ros_factory.so', world_path], output='screen'),
-        joint_state_publisher_node,
+    return LaunchDescription([
+        DeclareLaunchArgument(name='gui', default_value='True', description='Flag to enable joint_state_publisher_gui'),
```
There does not appear to be any users left of this argument

Amronos replied:
Fixed in [29276cb](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/29276cbcfe4fdf6c89f9f7b312f9e39b67dd0b64).

SteveMacenski commented on sam_bot_description/launch/display.launch.py:
```diff
@@ -1,60 +1,74 @@
+        DeclareLaunchArgument(name='model', default_value=default_model_path, description='Absolute path to robot urdf file'),
+        DeclareLaunchArgument(name='rvizconfig', default_value=default_rviz_config_path, description='Absolute path to rviz config file'),
+        DeclareLaunchArgument(name='use_sim_time', default_value='True', description='Flag to enable use_sim_time'),
+        ExecuteProcess(cmd=['gz', 'sim', '-g'], output='screen'),
```
There's a ros_gz launch file for this which includes some other useful things we should use instead 

Edit: how does this play with the GZServer() you already created?

Amronos replied:
Have a look at this: https://gazebosim.org/docs/harmonic/ros2_overview/#composition
Also, I think the launch file you are mentioning is [this](https://github.com/gazebosim/ros_gz/blob/ros2/ros_gz_sim/launch/gz_sim.launch.py.in). This launch file doesn't let you use composition and that's why I implemented what's done in [this](https://github.com/gazebosim/ros_gz/blob/ros2/ros_gz_sim/launch/ros_gz_sim.launch.py) launch file.
The ``ExecuteProcess(cmd=['gz', 'sim', '-g'], output='screen'),`` simply lets you start a Gazebo GUI, without the server which we are already launching before.

SteveMacenski reviewed (commented):
I'm a little unclear as to why there are `<gazebo>` tags and sensor plugin definitions in the URDF. I think the URDF is just the non-GZ things and the SDF contains the gazebo and sensor plugins that attach to frames defined in the URDF. I think things like the IMU gazebo stuff should be removed to the SDF and only the IMU frame is created / linked in the URDF (and similarly for other sensors, diff drive plugin, joint state publisher, etc). 

I don't think there needs to be / should be duplication. I'm not familiar with that much duplication in SDF and URDF

mikeferguson commented:
> I'm a little unclear as to why there are `<gazebo>` tags and sensor plugin definitions in the URDF

I haven't looked at this PR at all - but I will note that for a long time there have been gazebo tags in many URDFs - the URDF parser silently ignores them. The SDF has usually been more the description of the scene, not the robot. I'm not sure what the "best" practices are for new gazebo.

SteveMacenski commented:
For example, in `nav2_minimal_turtlebot_simulation` I (developed? found? modified? honestly I was doing so much that week its all a blur) put together a URDF that has no gazebo / sim things in it and a separate SDF that contains the Gazebo bits to decouple them
- https://github.com/ros-navigation/nav2_minimal_turtlebot_simulation/blob/main/nav2_minimal_tb3_sim/urdf/gz_waffle.sdf.xacro
- https://github.com/ros-navigation/nav2_minimal_turtlebot_simulation/blob/main/nav2_minimal_tb3_sim/urdf/turtlebot3_waffle.urdf

If we're going to bleed them together, then I think its probably better to just use a singular file then and save the duplication maintenance work. You can actually see that we did that for the TB4 portions of that repository that were modified and pared down from the iRobot / Clearpath stacks.

mikeferguson commented:
In the turtlebot example, isn't the SRDF got a lot of duplicated things from the URDF (for instance, just the base_link definition is basically fully duplicated - if one gets updated but not the other, things get out of date)? The workflow I've seen in the past (again, ROS 1) was to create [your base URDF, with no gazebo tags](https://github.com/ZebraDevs/fetch_ros/blob/ros1/fetch_description/robots/fetch.urdf)  and then use [xacro to include that and add the gazebo stuff in a more gazebo-specific urdf](https://github.com/ZebraDevs/fetch_gazebo/blob/gazebo9/fetch_gazebo/robots/fetch.gazebo.xacro).

I guess I'm assuming here that ROS 2 / new gazebo still has the tools to load a URDF/xacro directly (basically, it creates the SDF under the hood) - if that's gone, it's a moot point and users have to maintain separate SDF & URDF files.

SteveMacenski commented:
I think the intention is to long-term stop allowing or promoting users to use URDFs in this way. With this said, I forget entirely where I heard that so I could just be making that up :laughing:

mikeferguson commented:
> I think the intention is to long-term stop allowing or promoting users to use URDFs in this way due to the limiting nature of the URDF format w.r.t. SDF. With this said, I forget entirely where I heard that so I could just be making that up 😆

I mean, this is the team that built their own ROS-alternative for Gazebo internal comms, so... I feel like Gazebo keeps getting harder and harder to use, at the same time so many alternatives keep coming on the market. /rant

Amronos committed:
Implement review comments

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Add argument for using ekf

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Some fixes

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

SteveMacenski commented on sam_bot_description/config/tf_bridge_config.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+- ros_topic_name: "/tf"
```
Why in a separate file?

Edit: I see now, do we need this if we always use the EKF? If not, then feel free to remove. I just thought it was nice to show TF even if we don't use it in this demonstration since it'll be a common-ish want.

SteveMacenski replied:
Is this still needed to be seprate or included if you fixed the EKF?

SteveMacenski commented on sam_bot_description/launch/display.launch.py:
```diff
@@ -1,60 +1,107 @@
-    world_path=os.path.join(pkg_share, 'world/my_world.sdf')
-    
-    robot_state_publisher_node = launch_ros.actions.Node(
+    world_path = os.path.join(pkg_share, 'world/my_world.sdf')
```
```suggestion
    world_path = os.path.join(pkg_share, 'world', 'my_world.sdf')
```

Is the more correct way of doing this, for all around here as well :-) 

I know it was wrong before :wink:

SteveMacenski commented on sam_bot_description/launch/display.launch.py:
```diff
@@ -1,60 +1,107 @@
+        container_name='ros_gz_container',
+        create_own_container='False',
+        use_composition='True',
+        condition=UnlessCondition(LaunchConfiguration('ekf'))
```
Lets just recommend they use the EKF, why not :-)

SteveMacenski reviewed (commented):
You bring up an interesting point about the composition containers and GZ. That's pretty new. Want to open a PR for Nav2's bringup to do that as well? I think that would be nice to have the bridge composed there as well now that this is an option!

Amronos commented:
> You bring up an interesting point about the composition containers and GZ. That's pretty new. Want to open a PR for Nav2's bringup to do that as well? I think that would be nice to have the bridge composed there as well now that this is an option!

Sure!

Also about the EKF: the reason I have included the tf_bridge is that the ekf makes the robot rotate a bit too much on the ROS side (as can be seen in RViz). This doesn't happen when I give the EKF only the odom topic instead of both the odom and imu. I am not too sure of whether this is a problem with EKF or Gazebo.

Amronos committed:
Update sam_bot_description/launch/display.launch.py

Co-authored-by: Steve Macenski <stevenmacenski@gmail.com>
Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Add needed dependencies to package.xml

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Rename urdf_config.rviz to config.rviz

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Small argument description fix

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Fix ekf config

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Remove Gazebo specific things from URDF

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos commented:
@SteveMacenski It turns out that the problem with the robot rotating too much with ekf was that the ``ekf_config.yaml`` was set up incorrectly. I have updated it and the robot moves the same in Gazebo and RViz now. Should I remove the ``/tf`` topic bridge now?

I have also removed all Gazebo-specific things from the URDF according to our discussion at https://github.com/ros-navigation/docs.nav2.org/pull/618.

SteveMacenski commented on sam_bot_description/config/ekf.yaml:
```diff
@@ -33,15 +33,15 @@ ekf_filter_node:
                       false, false, false,
-                      false, false, false]
+                      false, false, true,
+                      true, true, false]
```
Set accelerations to `false` not true. IMU accelerations are notoriously not reliable for finite movements in robotics unless you have a crazy high end IMU

Amronos replied:
Done in [af90222](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/af902222d02f42e507f9d8c2ff023cacd36781fb).

SteveMacenski commented on sam_bot_description/src/description/sam_bot_description.sdf:
```diff
@@ -0,0 +1,422 @@
+      <sensor name="depth_camera" type="rgbd_camera">
+        <always_on>true</always_on>
+        <visualize>true</visualize>
+        <update_rate>30.0</update_rate>
```
Maybe reduce to 5hz? The laser is also at an unrealistically low rate, just for demo purposes. 30fps can be alot for non-gpu accelerated computers

Amronos replied:
Done in [af90222](https://github.com/ros-navigation/navigation2_tutorials/pull/102/commits/af902222d02f42e507f9d8c2ff023cacd36781fb).

SteveMacenski commented:
> Should I remove the /tf topic bridge now?

Yes please!

Amronos committed:
Implement review comments

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Small code style fix

Signed-off-by: Aarav Gupta <amronos275@gmail.com>

Amronos committed:
Merge branch 'rolling' into update-new-gazebo

Amronos commented:
@SteveMacenski you forgot to merge this. :-)

SteveMacenski commented:
Errr... umm.... nothing to see here :laughing:
